Fixes #547 Create Full Calendar Filter

tadean opened the pull request:
This PR modifies the `/calendar` page by adding a calendar selection widget for date navigation.

## Description
This PR adds an `az_calendar_filter` plugin that integrates the jqueryUI datepicker to provider an exposed calendar filter for navigation.

**The following items remain to be done:**

- [x] Convert ported JS to ES6
- [x] Adjust CSS styles for calendar 100% width on mobile in Bootstrap 4
- [x] Add in D7 functionality that highlights days with events (e.g. run view without filter and highlight cells that have events)

## Related Issue
#547 

## How Has This Been Tested?

- Verify the calendar widget is visible at `/calendar`
- Create some events
- Verify the events are visible when the appropriate days/months are selected

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

smayden commented:
I don't think this issue is introduced by this ticket (maybe other recent event work?), but I'm seeing the gray event boxes now run together on the /calendar page on desktop and mobile.

![image](https://user-images.githubusercontent.com/10873398/124833653-5d811980-df33-11eb-9427-d2ce61279b20.png)

They have spacing on the demo site:
https://live-azs-demo.pantheonsite.io/calendar

smayden commented:
Initial testing of calendar widget/navigation all works as expected. Thanks!

mmunro-ltrr commented:
Checked on a local Lando build, just the steps suggested here (verify the calendar widget is at `/calendar`, create some events, check they're only visible for the appropriate month-week-day selections). Confirmed @smayden 's observation about the grey event boxes running together. On the demo site the `node__content` class brings in a `margin-top: 10px;`, but I'm not seeing that here.

mmunro-ltrr commented:
Looked at a local Lando build of the 2.0.0-beta8 release, expecting it to show the same margins for the calendar entries as the demo site, but they were missing that margin-top as well. It does look like an unrelated issue.

tadean commented:
- Added event highlighting functionality on calendar cells
- Converted existing JS to ES6 and committed `yarn watch` converted copy
- Fixed width issues (barrio bug)

trackleft commented:
Tested this locally with demo content, it is very snappy and all of the filters seem to work as expected, outstanding work @tadean !

trackleft commented on modules/custom/az_event/js/az_calendar_filter.es6.js:
```diff
@@ -0,0 +1,205 @@
+      // We may have recieved new cell data. Refresh existing datepickers.
+      $(".az-calendar-filter-calendar").datepicker("refresh");
+
+      // Initialize calendar widget wrapper if needed.
```
Here is an example of es6 jQuery each in case it helps.  I know that one stumped me for a while. https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_paragraphs/az_paragraphs_text_media/js/az_paragraphs_az_text_media_youtube.es6.js#L23-L34

tadean commented on modules/custom/az_event/config/install/core.entity_form_display.node.az_event.default.yml:
```diff
@@ -89,9 +89,15 @@ content:
     weight: 11
     settings:
       show_extra: false
+      hide_date: false
```
I've added configuration to not hide the end date field on smart date, but the field is still hidden anyway. This may be a smart date issue. I may look for a patch or resolution to this in a followup.

joeparsons replied:
> I've added configuration to not hide the end date field on smart date, but the field is still hidden anyway. This may be a smart date issue. I may look for a patch or resolution to this in a followup.

Sounds like it'd be a good thing to create a Quickstart issue for, at least.

tadean commented:
All ES6 lint errors are now addressed and I believe this ready for review.

smayden reviewed (dismissed):
Looking good!

mmunro-ltrr reviewed (approved):
Tested on a local build, adding several Events & checking that the widget controls worked with them. No errors in the logs.

joeparsons commented:
@tadean Do you think we could consider replacing the jQuery UI Datepicker with something else (to reduce our reliance on jQuery) as a future improvement?

joeparsons commented on modules/custom/az_event/src/Plugin/views/filter/AZCalendarFilter.php:
```diff
@@ -0,0 +1,172 @@
+    return $operators;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
```
Nitpick.  It'd probably be good to add a description here since it doesn't look like this method is provided by a parent class (AFAICT).

joeparsons reviewed (approved):
Overall this looks great from a code perspective IMHO.  I did add one nitpick comment about the missing docblock for a method that doesn't actually appear to be inheriting/overriding a parent class.  I'll leave it to @tadean to decide whether that's worth addressing or not.

Really nice work.  Thanks!

tadean commented:
> @tadean Do you think we could consider replacing the jQuery UI Datepicker with something else (to reduce our reliance on jQuery) as a future improvement?

I think it'd be good to do so. Jquery UI is definitely showing its age at this point. I think we could target any javascript datapicker that gives callbacks that deals with which calendar cells get clicked on. The javascript is pretty separated from the Drupal plugin.
